internal/pkg/service/stream/storage/node/readernode/sliceupload/operator.go

```diff
@@ -0,0 +1,363 @@
+	etcd "go.etcd.io/etcd/client/v3"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/log"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop/op"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/servicectx"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/plugin"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/level/local/diskreader"
+	stagingConfig "github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/level/staging/config"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/model"
+	storageRepo "github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/model/repository"
+	statsCache "github.com/keboola/keboola-as-code/internal/pkg/service/stream/storage/statistics/cache"
+)
+
+type operator struct {
+	config     stagingConfig.OperatorConfig
+	clock      clock.Clock
+	logger     log.Logger
+	publicAPI  *keboola.PublicAPI
+	volumes    *diskreader.Volumes
+	statistics *statsCache.L1
+	storage    *storageRepo.Repository
+	plugins    *plugin.Plugins
+
+	slices         *etcdop.MirrorMap[model.Slice, model.SliceKey, *sliceData]
+	uploadedSlices *etcdop.MirrorMap[model.Slice, model.SliceKey, *uploadedSliceData]
+}
+
+type sliceData struct {
+	Slice *model.Slice
+
+	// Lock prevents parallel check of the same slice.
+	Lock *sync.Mutex
+
+	// Processed is true, if the entity has been modified.
+	// It prevents other processing. It takes a while for the watch stream to send updated state back.
+	Processed bool
+	// Uploading is true, if the entity is already uploaded to file.Provider stagin storage
+	Uploading bool
+}
+
+type uploadedSliceData struct {
+	filename string
+
+	// Lock prevents parallel check of the same slice.
+	Lock *sync.Mutex
+}
+
+type dependencies interface {
+	Logger() log.Logger
+	Clock() clock.Clock
+	Process() *servicectx.Process
+	KeboolaPublicAPI() *keboola.PublicAPI
+	Volumes() *diskreader.Volumes
+	StatisticsL1Cache() *statsCache.L1
+	StorageRepository() *storageRepo.Repository
+	Plugins() *plugin.Plugins
+}
+
+func Start(d dependencies, config stagingConfig.OperatorConfig) error {
+	var err error
+	o := &operator{
+		config:     config,
+		clock:      d.Clock(),
+		logger:     d.Logger().WithComponent("storage.node.operator.slice.upload"),
+		publicAPI:  d.KeboolaPublicAPI(),
+		volumes:    d.Volumes(),
+		statistics: d.StatisticsL1Cache(),
+		storage:    d.StorageRepository(),
+		plugins:    d.Plugins(),
+	}
+
+	// Graceful shutdown
+	wg := &sync.WaitGroup{}
+	ctx, cancel := context.WithCancel(context.Background())
+	d.Process().OnShutdown(func(_ context.Context) {
+		o.logger.Info(ctx, "closing slice upload operator")
+
+		// Stop mirroring
+		cancel()
+		wg.Wait()
+		ctx := context.WithoutCancel(ctx)
+		for {
+			remainingToUpload := 0
+			o.slices.ForEach(func(_ model.SliceKey, data *sliceData) (stop bool) {
```

Comment:
This probablly won't work due to closed `ctx` on `o.slices` level

Reply:
I don't understand this part, we can discuss it on a call.

Reply:
Based on discussion it was removed. Waitgroup should solve waiting for ongoing uploads of slices.